Remove ActivitySourceAdapter

utpilla opened the pull request:
Currently, the instrumentation libraries depend on ActivitySourceAdapter to set the activity sampling to comply with the TracerProvider sampler. This PR demonstrates how we could implement instrumentation libraries without using ActivitySourceAdapter. I have updated AspNet, AspNetCore, HttpClient, and gRPC client instrumentation libraries to work without ActivitySourceAdapter.

## Changes
- Removed `ActivitySourceAdapter`
- Removed `AddDiagnosticSourceInstrumentation`
- Added a public API called `AddLegacyOperationName` which creates an activity listener for the empty ActivitySourceName and 
maintains a list of operation names to be looked up against on ActivityStarted callback to decide if the legacy activity should be processed
- Added the sampling check in ActivityListener OnStarted to sample the activities generated by instrumentations correctly
- Added the instrumentation ActivitySource such as "OpenTelemetry.Instrumentation.AspNetCore" and "OpenTelemetry.Instrumentation.Http" to `sources` in `AddInstrumentation` method of `TracerProviderBuilderSdk.cs`. ActivitySource is updated from an empty string to the assembly name in `OnStartActivity` of the instrumentation listeners, therefore we need to add these assembly names as a Source to the provider so that ActivityStopped callback can be called.
- Moved the code to set ActivitySource and Activity Kind for instrumented activities to `ActivityInstrumentationHelper.cs` under OpenTelemetry\Internal which is used by both AspNetCore and HttpClient instrumentation libraries.



* [x] `CHANGELOG.md` updated for non-trivial changes

cijothomas commented on src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs:
```diff
@@ -92,7 +92,7 @@ public static TracerProvider Build(this TracerProviderBuilder tracerProviderBuil
         internal static TracerProviderBuilder AddDiagnosticSourceInstrumentation<TInstrumentation>(
             this TracerProviderBuilder tracerProviderBuilder,
-            Func<ActivitySourceAdapter, TInstrumentation> instrumentationFactory)
+            Func<TInstrumentation> instrumentationFactory)
```
`AddDiagnosticSourceInstrumentation` method can now be eliminated and simply use `AddInstrumentation`

utpilla replied:
Removed `AddDiagnosticSourceInstrumentation` method

cijothomas commented on src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs:
```diff
@@ -163,16 +163,18 @@ internal TracerProvider Build()
+            public readonly Func<object> Factory;
 
-            internal DiagnosticSourceInstrumentationFactory(string name, string version, Func<ActivitySourceAdapter, object> factory)
+            internal DiagnosticSourceInstrumentationFactory(string name, string version, Func<object> factory)
```
`DiagnosticSourceInstrumentationFactory` can be totally eliminated now.

utpilla replied:
Removed `DiagnosticSourceInstrumentationFactory`

cijothomas commented on src/OpenTelemetry/Trace/TracerProviderSdk.cs:
```diff
@@ -76,6 +75,24 @@ internal class TracerProviderSdk : TracerProvider
                 {
                     OpenTelemetrySdkEventSource.Log.ActivityStarted(activity);
 
+                    if (string.IsNullOrEmpty(activity.Source.Name))
```
we can have a list of Activity names, for which we need to run samplers, and only run this for those opt-in activity names. Running this for *all* the legacy activities would affect hot path performance.
(even with list of opt-in names, we need to run benchmarks to see if there is significant regression)

cijothomas replied:
Something like the below:

List<string> LegacyActivityNames; which is populated when instrumentation calls `AddLegacyActivity(string name)`.

if this list is empty, then bypass the whole section. This would be the case when someone is using pure ActivitySource based activities only.

utpilla replied:
I have added a HashSet<string> to store all the operation names for the DiagnosticSource instrumentations that were added.

cijothomas commented on src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs:
```diff
@@ -60,7 +60,7 @@ public static class TracerProviderBuilderExtensions
             configureHttpClientInstrumentationOptions?.Invoke(httpClientOptions);
 
-            builder.AddDiagnosticSourceInstrumentation((activitySource) => new HttpClientInstrumentation(activitySource, httpClientOptions));
+            builder.AddDiagnosticSourceInstrumentation(() => new HttpClientInstrumentation(httpClientOptions));
```
```suggestion
            builder.AddInstrumentation(() => new HttpClientInstrumentation(httpClientOptions));
```

cijothomas replied:
followed by something like builder.AddLegacyActivity("nameofactivity from httpclient")

utpilla replied:
I have created two new methods `AddInstrumentationActivitySource` and `AddLegacyActivityOperationName` which should be used in the `TracerProviderBuilderExtensions.cs` of the instrumentation projects such as AspNetCore and HttpClient.

`AddInstrumentationActivitySource` ensures that the ActivityListener's ActivityStop callback is called as the ActivitySource is updated by the OnStartActivity of instrumentation listeners.

`AddLegacyActivityOperationName` is used to limit the processing of activities in the activity listener to only those that are required.

cijothomas reviewed (changes requested):
Thanks! The overall approach looks good. Need to optimize the perf in hot path, so some bencmark results before/after would need to be seen.

cijothomas commented:
@alanwest @CodeBlanch please take a look. This would eliminate the need of ActivitySourceAdapater, by making the TracerProviderSdk itself being aware of legacy activities.

cijothomas commented on src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs:
```diff
@@ -129,55 +129,49 @@ internal TracerProviderBuilder AddProcessor(BaseProcessor<Activity> processor)
+        /// </summary>
+        /// <param name="instrumentationActivitySource">ActivitySource to add.</param>
+        /// <returns>Returns <see cref="TracerProviderBuilder"/> for chaining.</returns>
+        internal TracerProviderBuilder AddInstrumentationActivitySource(string instrumentationActivitySource)
```
we dont need this, as there is already `AddSource` and this is adding anything new not already offered by `AddSource`

utpilla replied:
Removed this.

cijothomas commented on src/OpenTelemetry/Trace/TracerProviderSdk.cs:
```diff
@@ -153,11 +164,14 @@ internal class TracerProviderSdk : TracerProvider
                     // Function which takes ActivitySource and returns true/false to indicate if it should be subscribed to
                     // or not.
-                    listener.ShouldListenTo = (activitySource) => regex.IsMatch(activitySource.Name);
+                    listener.ShouldListenTo = (activitySource) => string.IsNullOrEmpty(activitySource.Name) || regex.IsMatch(activitySource.Name);
```
We dont want to affect the regular flow here by subscribing to empty sources. Do that conditionally, based of if there atleast one entry in the hashset/other datastructure keeping list of legacy activity names.

utpilla replied:
Added the condition to check if the HashSet is not empty.

cijothomas commented on src/OpenTelemetry/Trace/TracerProviderSdk.cs:
```diff
@@ -33,34 +33,29 @@ internal class TracerProviderSdk : TracerProvider
+        private readonly HashSet<string> legacyActivityOperationNames;
         private BaseProcessor<Activity> processor;
+        private Action<Activity> getRequestedDataAction;
+        private bool needInstrumentations;
```
```suggestion
        private bool supportLegacyActivity;
```
```suggestion
        private bool needInstrumentations;
```

utpilla replied:
Renamed the field to supportLegacyActivity.

cijothomas commented on src/OpenTelemetry/Trace/TracerProviderSdk.cs:
```diff
@@ -76,6 +71,11 @@ internal class TracerProviderSdk : TracerProvider
                 {
                     OpenTelemetrySdkEventSource.Log.ActivityStarted(activity);
 
+                    if (this.needInstrumentations && legacyActivityOperationNames.Contains(activity.OperationName))
```
if the activity is created using `ActivitySource` and it *happens* to have same operation name as one of the legacyActivityOperationNames, then we'd be running sampler for them again here, which would be incorrect. We need to do this only for legacy activities - i.e the ones which has ActivitySource empty.

utpilla replied:
Yeah this is something I missed. There could be cases where ActivitySource creates an activity with the same operation name as the one created by DiagnosticSource. I have added a check for empty ActivitySource.

cijothomas commented on src/OpenTelemetry/Trace/TracerProviderSdk.cs:
```diff
@@ -76,6 +71,11 @@ internal class TracerProviderSdk : TracerProvider
                 {
                     OpenTelemetrySdkEventSource.Log.ActivityStarted(activity);
 
+                    if (this.supportLegacyActivity && legacyActivityOperationNames.Contains(activity.OperationName) && string.IsNullOrEmpty(activity.Source.Name))
```
How about the following. 
```
if (this.supportLegacyActivity && activity.Source.Name.Equals(string.Empty))
                    {
                        // We have a legacy activity in hand now.
                        // and it's name matches the one user is interested in
                        if (legacyActivityOperationNames.Contains(activity.OperationName))
                        {
                            this.getRequestedDataAction(activity);
                        }
                        else
                        {
                            // legacy activity doesn't match user configured list.
                            // no need to proceeding further.
                            return;
                        }
                    }
```

cijothomas replied:
similar in ActivityStopped.
```
if (this.supportLegacyActivity && activity.Source.Name.Equals(string.Empty))
                    {
                        // We have a legacy activity in hand now
                        // but not something user expressed interest in.
                        // Dont proceed any further.
                        if (!legacyActivityOperationNames.Contains(activity.OperationName))
                        {
                            return;
                        }
                    }
```

utpilla replied:
Fixed this.

CodeBlanch commented:
I like this approach! It kind of makes what I'm doing on #1761 impossible, but probably better to approach that through the runtime team anyway 😄

utpilla commented:
With this new approach, we have introduced a few conditional checks to the TracerProviderSdk hot path for processing activities. I ran some benchmark tests and found that this has increased the mean time to run for the Tracer Benchmark test by up to ~20ns.

I feel this is acceptable considering the convenience it brings by getting rid of ActivitySourceAdapter entirely.

Here are the benchmarking results:

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
Intel Core i7-9700 CPU 3.00GHz, 1 CPU, 8 logical and 8 physical cores
.NET Core SDK=5.0.103
  [Host]     : .NET Core 3.1.12 (CoreCLR 4.700.21.6504, CoreFX 4.700.21.6905), X64 RyuJIT
  DefaultJob : .NET Core 3.1.12 (CoreCLR 4.700.21.6504, CoreFX 4.700.21.6905), X64 RyuJIT


**With the new changes:**

|                    Method |      Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------- |----------:|---------:|---------:|-------:|------:|------:|----------:|
|                NoListener |  23.50 ns | 0.034 ns | 0.058 ns |      - |     - |     - |         - |
|    PropagationDataListner | 688.19 ns | 2.647 ns | 2.476 ns | 0.1078 |     - |     - |     680 B |
|            AllDataListner | 684.48 ns | 3.639 ns | 3.404 ns | 0.1078 |     - |     - |     680 B |
| AllDataAndRecordedListner | 687.80 ns | 2.414 ns | 2.140 ns | 0.1078 |     - |     - |     680 B |
|              OneProcessor | 716.29 ns | 1.442 ns | 1.349 ns | 0.1078 |     - |     - |     680 B |
|             TwoProcessors | 736.54 ns | 5.225 ns | 4.632 ns | 0.1078 |     - |     - |     680 B |
|           ThreeProcessors | 743.95 ns | 3.218 ns | 2.687 ns | 0.1078 |     - |     - |     680 B |
|        OneInstrumentation | 730.72 ns | 1.876 ns | 1.663 ns | 0.1078 |     - |     - |     680 B |
|       TwoInstrumentations | 708.67 ns | 1.221 ns | 1.020 ns | 0.1078 |     - |     - |     680 B |

I added two Benchmark tests called `OneInstrumentation` and `TwoInstrumentations`:
- `OneInstrumentation` adds `AspNetCoreInstrumentation` to the provider
- `TwoInstrumentations` adds `AspNetCoreInstrumentation` and `HttpClientInstrumentation` to the provider
Both of these tests also add one processor to the TracerProvider.



**Without these changes (main branch):**

|                    Method |      Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------- |----------:|---------:|---------:|-------:|------:|------:|----------:|
|                NoListener |  23.45 ns | 0.034 ns | 0.042 ns |      - |     - |     - |         - |
|    PropagationDataListner | 677.92 ns | 1.638 ns | 1.368 ns | 0.1078 |     - |     - |     680 B |
|            AllDataListner | 685.51 ns | 2.517 ns | 2.354 ns | 0.1078 |     - |     - |     680 B |
| AllDataAndRecordedListner | 684.66 ns | 1.621 ns | 1.354 ns | 0.1078 |     - |     - |     680 B |
|              OneProcessor | 720.67 ns | 1.455 ns | 1.361 ns | 0.1078 |     - |     - |     680 B |
|             TwoProcessors | 711.47 ns | 2.183 ns | 1.823 ns | 0.1078 |     - |     - |     680 B |
|           ThreeProcessors | 721.47 ns | 1.324 ns | 1.239 ns | 0.1078 |     - |     - |     680 B |

cijothomas commented on src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs:
```diff
@@ -42,7 +42,9 @@ public static class TracerProviderBuilderExtensions
 
-            builder.AddDiagnosticSourceInstrumentation((activitySource) => new AspNetInstrumentation(activitySource, aspnetOptions));
+            builder.AddInstrumentation(() => new AspNetInstrumentation(aspnetOptions));
+            builder.AddSource(typeof(AspNetInstrumentation).Assembly.GetName().Name);
```
Why is this needed?

utpilla replied:
We need to add the source as the instrumented activity's source will be updated from empty string to this in the OnStartActivityMethod. The ActivityStopped callback method checks if the activity's ActivitySource is present in the list of sources before calling Processor.OnEnd

cijothomas replied:
> The ActivityStopped callback method checks if the activity's ActivitySource is present in the list of sources before calling Processor.OnEnd

I dont see that check. ActivityStopped is simply checking if ActivitySource is empty or not.

utpilla replied:
Ahh yes! I should update the ActivityStopped callback to check if the ActivitySource was added to the provider. This would require us to have a private field for Sources which will have all the sources that were added to the provider using AddSource. Does that sound good?

utpilla replied:
The check for ActivitySource happens on `listener.ShouldListenTo`. If the source is not added then the ActivityStop callback will not be called `listener.ShouldListenTo` will return false for such an activity.

cijothomas commented on test/OpenTelemetry.Tests/Trace/TracerProviderSdkTest.cs:
```diff
@@ -263,38 +271,301 @@ public void TracerProvideSdkCreatesActivitySource()
+        [Fact]
+        public void SdkSamplesAndProcessesLegacyActivityWithRightConfig()
+        {
+            bool samplerCalled = false;
```
this is unused?

utpilla replied:
There is an Assert statement for it in the beginning of `StartAction` in the `TestActivityProcessor`.

cijothomas commented on src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs:
```diff
@@ -42,7 +42,9 @@ public static class TracerProviderBuilderExtensions
 
-            builder.AddDiagnosticSourceInstrumentation((activitySource) => new AspNetInstrumentation(activitySource, aspnetOptions));
+            builder.AddInstrumentation(() => new AspNetInstrumentation(aspnetOptions));
+            builder.AddSource(typeof(AspNetInstrumentation).Assembly.GetName().Name);
```
please move the sourcename to a constant. There is inconsistency now as this uses AspNetInstrumentation vs HttpInListener actually. (name is ultimately same, but lets do this consistently)

utpilla replied:
I have updated the code to use ActivitySourceName from the listener implementation for all four instrumentation libraries

cijothomas commented on src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs:
```diff
@@ -29,11 +29,10 @@ internal class AspNetCoreInstrumentation : IDisposable
-        /// <param name="activitySource">ActivitySource adapter instance.</param>
         /// <param name="options">Configuration options for ASP.NET Core instrumentation.</param>
-        public AspNetCoreInstrumentation(ActivitySourceAdapter activitySource, AspNetCoreInstrumentationOptions options)
+        public AspNetCoreInstrumentation(AspNetCoreInstrumentationOptions options)
```
I initially thought (inccoreclty) this is a breaking change! The public api analyzer is a gift!

utpilla replied:
Should we change this to internal?

cijothomas commented on src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs:
```diff
@@ -29,11 +29,10 @@ internal class AspNetCoreInstrumentation : IDisposable
+        public AspNetCoreInstrumentation(AspNetCoreInstrumentationOptions options)
         {
-            this.diagnosticSourceSubscriber = new DiagnosticSourceSubscriber(new HttpInListener("Microsoft.AspNetCore", options, activitySource), null);
+            this.diagnosticSourceSubscriber = new DiagnosticSourceSubscriber(new HttpInListener("Microsoft.AspNetCore", options), null);
```
there are lot of special names - lets move them to a common place where it can be commented and probbaly linked to aspnetcore repo source code.
(separate PR)

cijothomas commented on src/OpenTelemetry.Instrumentation.AspNet/TracerProviderBuilderExtensions.cs:
```diff
@@ -42,7 +42,9 @@ public static class TracerProviderBuilderExtensions
-            builder.AddDiagnosticSourceInstrumentation((activitySource) => new AspNetInstrumentation(activitySource, aspnetOptions));
+            builder.AddInstrumentation(() => new AspNetInstrumentation(aspnetOptions));
+            builder.AddSource(typeof(AspNetInstrumentation).Assembly.GetName().Name);
+            builder.AddLegacyActivityOperationName("Microsoft.AspNet.HttpReqIn");
```
dont we create sibling activities in asp.net also?

utpilla replied:
Thanks! I missed this one. I have added it.

cijothomas commented on src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs:
```diff
@@ -71,41 +71,31 @@ public static TracerProviderBuilder AddProcessor(this TracerProviderBuilder trac
-            this TracerProviderBuilder tracerProviderBuilder,
-            Func<ActivitySourceAdapter, TInstrumentation> instrumentationFactory)
-            where TInstrumentation : class
+        public static TracerProviderBuilder AddLegacyActivityOperationName(this TracerProviderBuilder tracerProviderBuilder, string operationName)
```
```suggestion
        public static TracerProviderBuilder AddLegacyActivity(this TracerProviderBuilder tracerProviderBuilder, string operationName)
```

utpilla replied:
I have updated the name to AddLegacyActivity and updated the public API docs as well.

cijothomas commented on src/OpenTelemetry/Trace/TracerProviderBuilderExtensions.cs:
```diff
@@ -71,41 +71,31 @@ public static TracerProviderBuilder AddProcessor(this TracerProviderBuilder trac
-        /// Adds a DiagnosticSource based instrumentation.
-        /// This is required for libraries which is already instrumented with
-        /// DiagnosticSource and Activity, without using ActivitySource.
+        /// Adds an activity listener for activities created the legacy way with the provided OperationName. The listener samples and processes these activities
```
/// Adds activity with given operation name to the list of subscribed activities. This is for legacy activities (i.e activities created without using ActivituSource API).

utpilla replied:
I have updated the description.

cijothomas commented on src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs:
```diff
@@ -129,55 +129,32 @@ internal TracerProviderBuilder AddProcessor(BaseProcessor<Activity> processor)
-        /// Adds a DiagnosticSource based instrumentation.
-        /// This is required for libraries which is already instrumented with
-        /// DiagnosticSource and Activity, without using ActivitySource.
+        /// Adds the OperationName of an activity created by DiagnosticSource instrumentation to the provider.
```
its misleading to say "created by DiagnosticSource instrumentation".

utpilla replied:
I have updated the description for this to match the public extension method's description.

cijothomas commented on test/Benchmarks/Trace/TraceBenchmarks.cs:
```diff
@@ -80,6 +82,21 @@ public TraceBenchmarks()
+            Sdk.CreateTracerProviderBuilder()
+                .SetSampler(new AlwaysOnSampler())
+                .AddSource(this.sourceWithOneInstrumentation.Name)
+                .AddAspNetCoreInstrumentation()
```
This benchmark has nothing to do with `AddAspNetCoreInstrumentation`. You can simply use the raw API directly - AddLegacyActivity("somename")

utpilla replied:
I have removed the instrumentations and used AddLegacyActivity.

cijothomas reviewed (approved):
LGTM.
Left a few comments, nothing blocking. Okay to have them as follow up PRs.

eddynaka commented on src/OpenTelemetry.Instrumentation.GrpcNetClient/TracerProviderBuilderExtensions.cs:
```diff
@@ -43,7 +44,10 @@ public static class TracerProviderBuilderExtensions
-            builder.AddDiagnosticSourceInstrumentation((activitySource) => new GrpcClientInstrumentation(activitySource, grpcOptions));
+            builder.AddInstrumentation(() => new GrpcClientInstrumentation(grpcOptions));
+            builder.AddSource(GrpcClientDiagnosticListener.ActivitySourceName);
+            builder.AddLegacyActivity("Grpc.Net.Client.GrpcOut");
```
nit: create as a const, just like aspnetcore

eddynaka commented on src/OpenTelemetry.Instrumentation.Http/TracerProviderBuilderExtensions.cs:
```diff
@@ -60,7 +58,9 @@ public static class TracerProviderBuilderExtensions
-            builder.AddDiagnosticSourceInstrumentation((activitySource) => new HttpClientInstrumentation(activitySource, httpClientOptions));
+            builder.AddInstrumentation(() => new HttpClientInstrumentation(httpClientOptions));
+            builder.AddSource(HttpHandlerDiagnosticListener.ActivitySourceName);
+            builder.AddLegacyActivity("System.Net.Http.HttpRequestOut");
```
nit: create const

eddynaka commented on src/OpenTelemetry/Internal/ActivityInstrumentationHelper.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace OpenTelemetry.Instrumentation
+{
+    internal static class ActivityInstrumentationHelper
```
should we compile the existing extension instead of duplicating it?

eddynaka commented on src/OpenTelemetry/Trace/TracerProviderBuilderSdk.cs:
```diff
@@ -129,55 +129,32 @@ internal TracerProviderBuilder AddProcessor(BaseProcessor<Activity> processor)
-        internal TracerProviderBuilder AddDiagnosticSourceInstrumentation<TInstrumentation>(
-            Func<ActivitySourceAdapter, TInstrumentation> instrumentationFactory)
-            where TInstrumentation : class
+        internal TracerProviderBuilder AddLegacyActivity(string operationName)
```
nit: we could change this to `params string[]`, we have many cases calling this twice.

cijothomas commented:
@open-telemetry/dotnet-approvers Please take a look. Will proceed to merge this in next few hours as 1st step in unblocking instrumentations in contrib repo. 
This is adding new public API, if there are feedbacks about the name, etc - we have opportunity to change it before the next stable release.

CodeBlanch commented on src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs:
```diff
@@ -99,6 +98,9 @@ public override void OnStartActivity(Activity activity, object payload)
                     newOne.Start();
+
+                    // Set IsAllDataRequested to false for the activity created by the framework to only export the sibling activity and not the framework activity
+                    activity.IsAllDataRequested = false;
```
@cijothomas Are we exporting these in 1.0? Could technically be breaking if so? I think it's worth doing but should be called out as a change. Might already be in changelog I haven't gotten that far in the diff yet 😄

cijothomas replied:
Instrumentation haven't reached 1.0, so okay to make breaking change. However, I dont think there is any breaking change (breaking behaviour). Previously we were exporting only one activity, which is the one we create, if we create one. The framework is not exported today, if we create sibling activity ourselves

CodeBlanch commented on src/OpenTelemetry/.publicApi/net452/PublicAPI.Unshipped.txt:
```diff
@@ -1 +1,2 @@
-static OpenTelemetry.Trace.TracerProviderExtensions.ForceFlush(this OpenTelemetry.Trace.TracerProvider provider, int timeoutMilliseconds = -1) -> bool
\ No newline at end of file
+static OpenTelemetry.Trace.TracerProviderBuilderExtensions.AddLegacyActivity(this OpenTelemetry.Trace.TracerProviderBuilder tracerProviderBuilder, string operationName) -> OpenTelemetry.Trace.TracerProviderBuilder
```
Not a big fan of this name. How about `AddLegacySource`? Given we have `AddSource` already.

cijothomas replied:
I am okay with `AddLegacySource`

cijothomas replied:
we have opportunity to rename this until 1.1 stable is out, so i'd prefer to not block this pr.

CodeBlanch replied:
Fine by me, I'll stamp as approved.
